### PR TITLE
feat: support circular refs

### PR DESCRIPTION
**Pull request by pytlesk4**
Addresses https://github.com/stoplightio/prism/issues/1456

Blocked By https://github.com/stoplightio/json/pull/90

**Summary**

- switch to our forked json schema ref parser
- bundle http operations
- validate circular data properly
- dynamically generate data for circular structures
- tests

**Checklist**

- The basics
  - [x] I tested these changes manually in my local or dev environment
- Tests
  - [x] Added or updated
  - [ ] N/A
- Event Tracking
  - [ ] I added event tracking and followed the event tracking guidelines
  - [x] N/A
- Error Reporting
  - [ ] I reported errors and followed the error reporting guidelines
  - [x] N/A

### CODE REVIEW

File: packages/cli/src/operations.ts [hunk 1]
```diff
@@ -29,3 +43,38 @@ function isOpenAPI3(document: unknown): document is OpenAPIObject {
...
+
+// TODO: remove, need https://github.com/stoplightio/json/pull/90
+export const decycle = (obj: unknown, replacer?: (value: any) => any) => {
+  const objs = new WeakMap<object, string>();
+  return (function derez(value: any, path: string[]) {
+    // The new object or array
+    let curObj: any;
+
+    // If a replacer function was provided, then call it to get a replacement value.
+    if (replacer) value = replacer(value);
```
Comment (radzserg) [inline packages/cli/src/operations.ts:55]: we don't path replacer, why do we need it?

File: packages/cli/src/operations.ts [hunk 2]
```diff
@@ -29,3 +43,38 @@ function isOpenAPI3(document: unknown): document is OpenAPIObject {
 function isPostmanCollection(document: unknown): document is CollectionDefinition {
   return Array.isArray(get(document, 'item')) && get(document, 'info.name');
 }
+
+// TODO: remove, need https://github.com/stoplightio/json/pull/90
+export const decycle = (obj: unknown, replacer?: (value: any) => any) => {
+  const objs = new WeakMap<object, string>();
+  return (function derez(value: any, path: string[]) {
```
Comment (radzserg) [inline packages/cli/src/operations.ts:50]: what's `derez`? Let's avoid reductions in a favor of meaningful names.

File: packages/cli/src/operations.ts [hunk 3]
```diff
@@ -29,3 +43,38 @@ function isOpenAPI3(document: unknown): document is OpenAPIObject {
 function isPostmanCollection(document: unknown): document is CollectionDefinition {
   return Array.isArray(get(document, 'item')) && get(document, 'info.name');
 }
+
+// TODO: remove, need https://github.com/stoplightio/json/pull/90
+export const decycle = (obj: unknown, replacer?: (value: any) => any) => {
+  const objs = new WeakMap<object, string>();
```
Comment (radzserg) [inline packages/cli/src/operations.ts:49]: Let's make it in fp-ts style.

**Review by radzserg — DISMISSED** [review]
Let's build derez in `fp-ts` style

File: packages/cli/src/operations.ts (hunk 3 above)
Comment (pytlesk4) [reply packages/cli/src/operations.ts:49]: This function is going to be replaced, with the json package one.

File: packages/cli/src/operations.ts (hunk 2 above)
Comment (pytlesk4) [reply packages/cli/src/operations.ts:50]: This function is going to be replaced, with the json package one.

File: packages/cli/src/operations.ts (hunk 1 above)
Comment (pytlesk4) [reply packages/cli/src/operations.ts:55]: This function is going to be replaced, with the json package one.

File: packages/cli/src/operations.ts (hunk 2 above)
Comment (radzserg) [reply packages/cli/src/operations.ts:50]: then what are we waiting for :) ? 
https://github.com/stoplightio/json/pull/90 is merged.

**Review by chohmann — COMMENTED** [review]
Tested with the following spec:
```
openapi: 3.0.0
info:
  title: Test API
  version: '1.0'
servers:
  - url: 'http://localhost:3000'
paths:
  '/dog/{id}':
    parameters:
      - schema:
          type: string
        name: id
        in: path
        required: true
    get:
      summary: Get A Dog
      tags: []
      responses:
        '200':
          description: OK
          content:
            application/json:
              schema:
                $ref: '#/components/schemas/Dog'
components:
  schemas:
    Dog:
      title: Dog
      type: object
      properties:
        id:
          type: number
        name:
          type: string
        breed:
          type: string
        parent:
          $ref: '#/components/schemas/Dog'
```

And when the response from that endpoint:
```
{
   "id":0,
   "name":"string",
   "breed":"string",
   "parent":{
      "id":0,
      "name":"string",
      "breed":"string",
      "parent":{}
   }
}
```

So working well for me.

Just a few comments/questions.

File: packages/cli/src/operations.ts [hunk 4]
```diff
@@ -2,20 +2,34 @@ import { transformOas3Operations } from '@stoplight/http-spec/oas3/operation';
...
-  if (isPostmanCollection(result)) return transformPostmanCollectionOperations(result);
+  let operations: IHttpOperation[] = [];
+  if (isOpenAPI2(result)) operations = transformOas2Operations(result);
+  else if (isOpenAPI3(result)) operations = transformOas3Operations(result);
+  else if (isPostmanCollection(result)) operations = transformPostmanCollectionOperations(result);
+  else throw new Error('Unsupported document format');
 
-  throw new Error('Unsupported document format');
+  operations.forEach((op, i, ops) => {
+    // TODO: should we add a __bundled__ property to stoplight/types for http operations?
```
[the remark this answers is not in the document]
Comment (pytlesk4) [reply packages/cli/src/operations.ts:22]: Yes, this is something we can do now. We don't have to use __bundled__ either, maybe we can add a bundled property to http operation?

Then we would just need to set the bundle root to whatever we decide for the name of the property on http operation. 

So if we choose `bundled`:

```ts
ops[i] = bundleTarget({      
  document:  {        
     ...result,
     __target__: op,
  }, 
  path: '#/__target__',
  bundleRoot: '#/bundled',
}); 
```

File: packages/http-server/src/__tests__/body-params-validation.spec.ts [hunk 5]
```diff
@@ -251,6 +251,67 @@ describe('body params validation', () => {
...
+            },
+            headers: [],
+            query: [],
+            cookie: [],
+            path: [],
+          },
+          tags: [],
+          security: [],
+          // @ts-ignore
+          ['__bundled__']: {
```
[the remark this answers is not in the document]
Comment (pytlesk4) [reply packages/http-server/src/__tests__/body-params-validation.spec.ts:297]: Yeah

File: packages/http-server/src/__tests__/server.oas.spec.ts [hunk 6]
```diff
@@ -168,7 +168,9 @@ describe.each([[oas2File], [oas3File]])('server %s', file => {
   });
 
   it('will return requested response using the __code property', async () => {
-    const response = await makeRequest('/pets/123?__code=404');
+    // TODO: Figure out why dynamic being enabled on the server returns a 200, instead of a 404.
+    // const response = await makeRequest('/pets/123?__code=404');
+    const response = await makeRequest('/pets/123?__code=404&__dynamic=false');
```
[the remark this answers is not in the document]
Comment (pytlesk4) [reply packages/http-server/src/__tests__/server.oas.spec.ts:173]: Will do!

File: packages/cli/src/operations.ts (hunk 4 above)
[the remark this answers is not in the document]
Comment (chohmann) [reply packages/cli/src/operations.ts:22]: what do you propose?

And should we make a change with this PR or a follow up PR?

File: packages/http-server/src/__tests__/server.oas.spec.ts (hunk 6 above)
[the remark this answers is not in the document]
Comment (pytlesk4) [reply packages/http-server/src/__tests__/server.oas.spec.ts]: Happens on master.

File: packages/cli/src/operations.ts (hunk 4 above)
[the remark this answers is not in the document]
Comment (pytlesk4) [reply packages/cli/src/operations.ts]: Issue: https://github.com/stoplightio/platform-internal/issues/6494 will add more context

File: packages/http-server/src/__tests__/server.oas.spec.ts (hunk 6 above)
[the remark this answers is not in the document]
Comment (pytlesk4) [reply packages/http-server/src/__tests__/server.oas.spec.ts]: Issue https://github.com/stoplightio/prism/issues/1836

**Review by chohmann — APPROVED** [review]
LGTM!

**Pull request merged**